core/src/main/java/org/infinispan/interceptors/locking/AbstractLockingInterceptor.java

```diff
@@ -146,23 +149,33 @@ protected final Throwable cleanLocksAndRethrow(InvocationContext ctx, Throwable
       return te;
    }
 
-   protected final boolean shouldLock(Object key, FlagAffectedCommand command) {
-      if (hasSkipLocking(command))
-         return false;
-      if (cacheConfiguration.clustering().cacheMode() == CacheMode.LOCAL)
-         return true;
-      boolean shouldLock = cdl.localNodeIsPrimaryOwner(key);
+   protected final long getLockTimeoutMillis(LocalFlagAffectedCommand command) {
+      return command.hasFlag(Flag.ZERO_LOCK_ACQUISITION_TIMEOUT) ? 0 : cacheConfiguration.locking().lockAcquisitionTimeout();
+   }
+
+   protected final boolean shouldLockKey(Object key) {
+      //only the primary owner acquires the lock.
+      boolean shouldLock = LockUtil.getLockOwnership(key, cdl) == LockUtil.LockOwnership.PRIMARY;
       getLog().tracef("Are (%s) we the lock owners for key '%s'? %s", cdl.getAddress(), key, shouldLock);
       return shouldLock;
    }
 
-   protected final void lockKey(InvocationContext ctx, DataWriteCommand command) throws InterruptedException {
-      boolean skipLocking = hasSkipLocking(command);
-      long lockTimeout = getLockAcquisitionTimeout(command, skipLocking);
-      lockKey(ctx, command.getKey(), lockTimeout, skipLocking);
+   protected final void lockAndRecord(InvocationContext context, Object key, long timeout) throws InterruptedException {
+      lockManager.acquireLock(context, key, timeout, false);
    }
 
-   protected final void lockKey(InvocationContext ctx, Object key, long timeoutMillis, boolean skipLocking) throws InterruptedException {
-      lockManager.acquireLockNoCheck(ctx, key, timeoutMillis, skipLocking);
+   protected final void lockAllAndRecord(InvocationContext context, Collection<?> keys, long timeout) throws InterruptedException {
+      for (Object key : keys) {
+         lockManager.acquireLock(context, key, timeout, false);
+      }
+   }
+
+   protected final Collection<Object> filterKeysToLock(Collection<Object> keys) {
```

Comment:
I have doubts because of the following:
* I don't like to have ```lock*``` and ```lockAll*``` variants. But in tx, you usually get multiple keys and in non-tx, you get a single key. If I could have a method that works fast enough for both scenarios, it would be great :)
* I don't wanna add a ```lockStream*``` variant. It would be good for PutMap but in tx I need to filter the keys by primary and backup owners (so I still need the ```lockAll*```). I'm doing the filtering here to avoid iterate over keys multiple times: https://github.com/infinispan/infinispan/pull/3594/files#diff-399714d17ac7a50a5a4e6b9f74620076R34
* If I use only ```lockStream*```, I could wrap the single key in a ```Stream``` and I could create two ```Stream```s for the scenario above, but it looks expensive.

I'm open to suggestions and yours expertise :D